Source/Dafny/Compiler-java.cs

```diff
@@ -2030,6 +2030,10 @@ private class GenericArrayElementLvalue : ILvalue {
             Contract.Assert(false);
             throw new cce.UnreachableException();
         }
+      } else if (arg.Type.AsArrayType != null) {
+        wr.Write("java.util.Arrays.toString(");
```

Comment:
This is not right. `java.util.Arrays.toString` prints out some representation of the _contents_ of the array, which is not what we want. Dafny prints references (which includes arrays) without following the data being referenced (which would cause problems with cyclic structures). Other back-ends print the name of the type of the reference. Previously, I guess the Java backend had printed some representation of the pointer itself, which cause this test to be nondeterministic.

My suggestion is to undo this printing change in Compiler-java.cs. I think it's fine behavior for the printing to output the reference itself (which is nondeterministic). Instead, we should change the test case, so that it doesn't rely on nondeterministic output. For example, the line

    print cb.y, " ", cb.k, " ", cb.l, "\n";

in the test's `Main` method can be changed to

    print cb.y[..], " ", cb.k[..], " ", cb.l[..], "\n";

Reply:
OK - done. There should probably be a separate issue to review conversion to string for all kinds of values and languages.